shared-module/displayio/Bitmap.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2018 Scott Shawcroft for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/displayio/Bitmap.h"
+
+#include <string.h>
+
+#include "py/runtime.h"
+
+void common_hal_displayio_bitmap_construct(displayio_bitmap_t *self, uint32_t width,
+    uint32_t height, uint32_t value_size) {
+    uint32_t row_width = width * value_size;
+    // word align
+    if (row_width % 32 != 0) {
+        self->stride = (row_width / 32 + 1);
+    } else {
+        self->stride = row_width / 32;
+    }
+    self->width = width;
+    self->height = height;
+    self->data = m_malloc(self->stride * height * sizeof(uint32_t), false);
+
+    self->bits_per_value = value_size;
+
+    self->x_shift = 0;
+    uint32_t power_of_two = 1;
+    while (power_of_two < 32 / value_size ) {
+        self->x_shift++;
+        power_of_two <<= 1;
+    }
+    self->x_mask = (1 << self->x_shift) - 1; // Used as a modulus on the x value
```

Comment:
I'm confused as to what this is computing, and what `x_shift` means in terms of extracting packed n-bit fields. If `bits_per_value` == 1, `x_shift` == 5; if bpv = 2 or 3, `x_shift` is 4, which is not intuitive to me. Computing `ceil(log2(32 // bits_value))`. Could you just comment this a bit, and check the numbers? (see below)

```
def x_shift(bits_per_value):
    shift = 0
    power_of_two = 1
    while power_of_two < 32 // bits_per_value:
        shift += 1
        power_of_two <<= 1
    return (shift, power_of_two)

>>> for i in range(1,32):
...     print(i, x_shift(i))

1 (5, 32)
2 (4, 16)
3 (4, 16)
4 (3, 8)
5 (3, 8)
6 (3, 8)
7 (2, 4)
8 (2, 4)
9 (2, 4)
10 (2, 4)
11 (1, 2)
12 (1, 2)
13 (1, 2)
14 (1, 2)
15 (1, 2)
16 (1, 2)
17 (0, 1)
18 (0, 1)
19 (0, 1)
20 (0, 1)
21 (0, 1)
22 (0, 1)
23 (0, 1)
24 (0, 1)
25 (0, 1)
26 (0, 1)
27 (0, 1)
28 (0, 1)
29 (0, 1)
30 (0, 1)
31 (0, 1)

Reply:
Oh, `x_shift` is a word index, I thought it was a bit index. OK, that makes a lot more sense to me.